### PR TITLE
#4 size estimator

### LINKED ISSUE #4: Implement `RecordSizer.rowSize` method
## Background
To be able to get record size from `DataFrame` the row size is needed.

## Feature
Implement `RecordSizer.rowSize` method

### PR DESCRIPTION
**Pull request by AdrianOlosutean**
Closes #4

### CODE REVIEW

**Review by dk1844 — DISMISSED** [review]
`SizeEstimator.estimate` 😮 Wow, just wow. Nicely implemented 😆 

(LGTM, just read the code)

(The current link to the original article is probably https://www.infoworld.com/article/2077408/sizeof-for-java.html)

**Comment by AdrianOlosutean** [comment]
I was initially thinking to use Java Instrumentation, but then discovered this `SizeEstimator` which probably fits the purpose. I don't have very deep knowledge of these, so it's open to debate

**Comment by benedeki** [comment]
Looks great. Could you do some tests/research if the estimator gives reasonable/sane numbers? 👼

**Comment by AdrianOlosutean** [comment]
After reading a bit on the topic I realized that SizeEstimator is not too reliable for estimating `GenericRowWithSchema` since it includes the schema object as well, so I implemented some logic to circumvent this, but it could be too slow for computation and probably use too much memory. This remains to be seen

File: src/main/scala/za/co/absa/spark_partition_sizing/RowSizer.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema
+import org.apache.spark.util.SizeEstimator
+import za.co.absa.spark_partition_sizing.types.ByteSize
+
+
+object RowSizer {
+
+  def rowSize(row: Row)(implicit fieldNames: Seq[String]): ByteSize = {
```
Comment (benedeki) [inline src/main/scala/za/co/absa/spark_partition_sizing/RowSizer.scala:27]: Nice to have an abstraction, where the implementation can change.
But I would wary to have such a common type as `Seq[String]` as `implicit` parameter, it's borderline imho.

File: src/main/scala/za/co/absa/spark_partition_sizing/RowSizer.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    rowEstimate(row)
+  }
+
+  private def rowEstimate(row: Row)(implicit fieldNames: Seq[String]): Long = {
+    row match {
+      case r: GenericRowWithSchema => {
+        val values = r.getValuesMap[AnyRef](fieldNames).values
+        values.map {
+          case struct: GenericRowWithSchema => rowEstimate(struct)(struct.schema.fields.map(_.name))
```
Comment (benedeki) [inline src/main/scala/za/co/absa/spark_partition_sizing/RowSizer.scala:36]: Recursive. Would turn in `@tailrec`

Comment (AdrianOlosutean) [reply src/main/scala/za/co/absa/spark_partition_sizing/RowSizer.scala:36]: Cannot be applied here because the recursive call is not in the tail position

Comment (benedeki) [reply src/main/scala/za/co/absa/spark_partition_sizing/RowSizer.scala:36]: I believe the code can be altered to fulfill the condition. 😉 Programming exercise 🤓

Comment (AdrianOlosutean) [reply src/main/scala/za/co/absa/spark_partition_sizing/RowSizer.scala:36]: I think the problem here is the sum. I also tried something like:
`
if (!row.isInstanceOf[GenericRowWithSchema]) SizeEstimator.estimate(row)
    else {
      val values = row.getValuesMap[AnyRef](fieldNames).values
      values.map(r => {
        val ro = r.asInstanceOf[GenericRowWithSchema]
        rowEstimate(ro, ro.schema.fields.map(_.name))
      }).sum
    }
`

Comment (benedeki) [reply src/main/scala/za/co/absa/spark_partition_sizing/RowSizer.scala:36]: You are right `@tailrec` would be harder to implement (despite not impossible).
I added the suggestion though, that simplifies the code, removes the double case and most of all gets rid if the `implicit`

File: src/main/scala/za/co/absa/spark_partition_sizing/RowSizer.scala [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+          case a => SizeEstimator.estimate(a)
+        }.sum
+      }
+      case _ => SizeEstimator.estimate(row)
+    }
```
Comment (benedeki) [inline src/main/scala/za/co/absa/spark_partition_sizing/RowSizer.scala:41]:
```suggestion
    rowEstimate(row, 0)
  }

  private def rowEstimate(item: AnyRef, accumulator: ByteSize): ByteSize = {
    item match {
      case row: GenericRowWithSchema =>
        val fieldNames: Seq[String] = row.schema.fields.map(_.name)
        val values = row.getValuesMap[AnyRef](fieldNames).values
        values.foldLeft(accumulator) {(acc, item) => rowEstimate(item, acc)}
      case _ => accumulator + SizeEstimator.estimate(item)
    }
  }
```

**Review by benedeki — APPROVED** [review]

**Pull request merged**